siva/repository.go

```diff
@@ -85,50 +75,48 @@ func (r *Repository) Mode() borges.Mode {
 
 // Commit implements borges.Repository interface.
 func (r *Repository) Commit() error {
-	if !r.transactional {
-		return borges.ErrNonTransactional.New()
-	}
-
-	if r.mode != borges.RWMode {
-		return nil
-	}
-
 	r.mu.Lock()
 	defer r.mu.Unlock()
 
 	if r.closed {
 		return ErrRepoAlreadyClosed.New(r.id)
 	}
 
-	err := r.fs.Sync()
-	if err != nil {
-		return err
+	if !r.transactional {
+		return borges.ErrNonTransactional.New()
+	}
+
+	defer func() { r.closed = true }()
+
+	sto, ok := r.s.(*Storage)
+	if ok {
+		err := sto.Commit()
+		if err != nil {
+			return err
+		}
 	}
 
-	r.closed = true
 	return r.location.Commit(r.mode)
 }
 
 // Close implements borges.Repository interface.
 func (r *Repository) Close() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
 	if r.closed {
 		return ErrRepoAlreadyClosed.New(r.id)
 	}
+	defer func() { r.closed = true }()
```

Comment:
Same here about marking the repository as closed if `sto.Close()` fails, is it fine?